dapp/contracts/loans/InvestorLedger.sol

```diff
@@ -1,158 +1,452 @@
+            require(investorData.amountInvested + investmentAmount > investorData.amountInvested);
+            require(ledger.totalAmountInvested + investmentAmount > ledger.totalAmountInvested);
+            investorData.amountInvested += investmentAmount;
+            ledger.totalAmountInvested += investmentAmount;
+            // Did we just gather all investments required?
+            if (ledger.totalAmountInvested == ledger.amountWanted) {
+                newState(ledger, State.Payback);
+                ledger.paybackDeadline = timestamp + ledger.paybackDelta;
+            }
+            // Transfer the investment to the receiving contract.
+            InvestmentSent(this, investor, investmentAmount);
+            require(
+                ledger.loanToken.transferFrom(
+                    investor,
+                    this,
+                    investmentAmount
+                )
+            );
+            // If this was the last required invesment, transfer all the
+            // invested money to the borrower.
+            if (ledger.totalAmountInvested == ledger.amountWanted) {
+                require(
+                    ledger.loanToken.transfer(
+                        ledger.borrower,
+                        ledger.amountWanted
+                    )
+                );
+            }
+        }
     }
-    
-    function calculateInterest(Ledger storage account, uint256 investment) private constant returns (uint256 interest) {
-        return investment * account.interestPermil / PERMIL;
+
+    // paybackProcess prforms processing within the Payback state of the loan
+    // FSM and possibly advances to its' next state. It can:
+    // - collect a full payback from the borrower and advance to the Paidback
+    //   state
+    // - if timeout is reached, advance to the Paidback state
+    function paybackProcess(Ledger storage ledger, address caller) public {
+        require(ledger.state == State.Payback);
+
+        if (!processTimeouts(ledger)) {
+            return;
+        }
+        if (caller != ledger.borrower) {
+            return;
+        }
+
+        // Gather payback, if possible.
+        uint256 payback = paybackRequired(ledger);
+        if (ledger.loanToken.allowance(ledger.borrower, this) >= payback) {
+            newState(ledger, State.Paidback);
+            require(
+                ledger.loanToken.transferFrom(
+                    ledger.borrower,
+                    this,
+                    payback
+                )
+            );
+        }
     }
 
-    struct Ledger {
-        // Config
-        IToken collateralToken;
-        IToken loanToken;
+    // canWithdrawLoanToken performs processing to see if a given address can
+    // withdraw loan tokens in one of the final states (Paidback, Canceled,
+    // Defaulted)
+    function canWithdrawLoanToken(Ledger storage ledger, address caller) public returns (uint256 _amount) {
+        processTimeouts(ledger);
```

Comment:
Because I made processTimeouts able to be called in a two-fold manner:
 - call it straight away to possibly process timeouts and then check state (the way it's called here)
 - first require() state then check the result of processTimeouts() to see if it has changed.

Which isn't great, so we should get rid of one of them. I guess the first/modifier approach is better, even if we don't have modifier syntax that we can use here..? WDYT?

Reply:
you can also have `requireState()` modifier btw

Reply:
I feel like using a modifier for requireState() is overengineering and doesn't either create more readable or elegant code. `require(ledger.state == Foo)` seems easy enough to understand to not warrant abstracting away.